pkg/config/expand.go

```diff
@@ -412,6 +412,59 @@ type varContext struct {
 	blueprint  Blueprint
 }
 
+/*
+A module reference is made by the use keyword and is subject to IGC constraints
+of references (ordering, explicitness). It has the following fields:
+  - ID: a module ID
+  - GroupID: the deployment group in which the module is *expected* to be found
+  - SourceGroupID: the deployment group from which the reference is made
+  - ExplicitInterGroup: a boolean value indicating whether the user made a
+    reference that superficially appears to be intergroup (i.e. they used
+    an explicit GroupID that differs from the context's GroupID)
+*/
+type modReference struct {
+	ID                 string
+	GroupID            string
+	SourceGroupID      string
+	ExplicitInterGroup bool
```

Comment:
It became clear that the only thing we "really" care about enforcing in a reference is when the reference is both explicit and intergroup, thus only bother tracking both at the same time.

- if the reference is intergroup (as determined by where the modules are actually found), then we care that the reference was explicitly intergroup
- if the reference is intragroup (as determined by where the modules are actually found), we do not care about explicitness alone, but if it was explicitly intergroup, then we know it's wrong

Maybe this will fall out of fixing the test for "ToGroupID" in the intergroup case. I will consider it.

Reply:
I think a case could be made to use `ExplicitInterGroup` but addressing some of the other concerns in the PR made that case a littler weaker. I have modified it to be Explicit. This will be reflected in the discussed mini-patch to varReference to align implementations.